baseServer.go

```diff
@@ -0,0 +1,216 @@
+	agentIdSet map[uuid.UUID]struct{}
+	// waitgroup for message synchronisation
+	listeningWaitGroup *sync.WaitGroup
+	// map of uuid ->struct{}{} which stores the ids of agents which have stopped messaging
+	agentStoppedTalkingMap map[uuid.UUID]struct{}
+	//channel a server goroutine will send to in order to signal messaging completion
+	messagingFinished      chan struct{}
+	// duration after which messaging phase forcefully ends during rounds
+	turnTimeout time.Duration
+	// run single round
+	roundRunner RoundRunner
+	//iterations
+	iterations int
+	//turns
+	turns int
+	//lock which stops race conditions when updating agentStoppedTalkingMap 
+	agentMapRWMutex sync.RWMutex
+	//stops multiple sends to messagingFinished during a round
+	doneChannelOnce sync.Once
+}
+
+func (server *BaseServer[T]) HandleStartOfTurn(iter, round int) {
+	server.doneChannelOnce = sync.Once{}
+	server.messagingFinished = make(chan struct{}, 1)
+	fmt.Printf("Iteration %d, Round %d starting...\n", iter, round)
+
+}
+
+func (serv *BaseServer[T]) endAgentListeningSession() bool {
+
+	ctx, cancel := context.WithTimeout(context.Background(), serv.turnTimeout)
+	defer cancel()
+	select {
+	case <-serv.messagingFinished:
+		serv.agentStoppedTalkingMap = make(map[uuid.UUID]struct{})
+		return true
+
+	case <-ctx.Done():
+		serv.agentStoppedTalkingMap = make(map[uuid.UUID]struct{})
+		return false
+	}
+}
+
+func (server *BaseServer[T]) HandleEndOfTurn(iter, round int) {
+	server.endAgentListeningSession()
+	fmt.Printf("Iteration %d, Round %d finished.\n", iter, round)
+}
+
+func (server *BaseServer[T]) RunAgentLoop() {}
+
+func (server *BaseServer[T]) SendMessage(msg IMessage[T], receivers []uuid.UUID) {
+	for _, receiver := range receivers {
+		go msg.InvokeMessageHandler(server.agentMap[receiver])
+	}
+
+}
+
+func (serv *BaseServer[T]) AddAgent(agent T) {
+	serv.agentMap[agent.GetID()] = agent
+	serv.agentIdSet[agent.GetID()] = struct{}{}
+}
+
+func (serv *BaseServer[T]) ViewAgentIdSet() map[uuid.UUID]struct{} {
+	return serv.agentIdSet
+}
+
+func (serv *BaseServer[T]) AccessAgentByID(id uuid.UUID) T {
+	return serv.agentMap[id]
+}
+
+func (serv *BaseServer[T]) Initialise() {}
+
+func (serv *BaseServer[T]) Start() {
+	serv.checkHandler()
+	turns := serv.turns
+	iterations := serv.iterations
+	for i := 0; i < iterations; i++ {
+		for j := 0; j < turns; j++ {
+			serv.HandleStartOfTurn(i+1, j+1)
+			serv.roundRunner.RunTurn()
+			serv.HandleEndOfTurn(i+1, j+1)
+		}
+	}
+}
+
+func (serv *BaseServer[T]) GetAgentMap() map[uuid.UUID]T {
+	return serv.agentMap
+}
+
+func (serv *BaseServer[T]) agentStoppedTalking(id uuid.UUID) {
+	serv.agentMapRWMutex.Lock()
+	fmt.Println("sending stop talking request,id:", id)
+	serv.agentStoppedTalkingMap[id] = struct{}{}
+
+	if len(serv.agentStoppedTalkingMap) == len(serv.agentMap) {
+		serv.doneChannelOnce.Do(func() {
+			select {
+			case serv.messagingFinished <- struct{}{}:
+				close(serv.messagingFinished)
+			default:
```

Comment:
this is dangerous. we need this send to be *blocking*, to stop it from potentially being undelivered. If the default executes, we're screwed: i.e., if the first case statement fails, then it fails for the entire round

Reply:
i changed this but i also had to add a new message handler to messages for synchronous sending, as with one thread this will block forever (main thread sends but no thread exists to receive).